homeassistant/components/knx/number.py

```diff
@@ -0,0 +1,122 @@
+        return
+
+    platform_config = discovery_info["platform_config"]
+    xknx: XKNX = hass.data[DOMAIN].xknx
+
+    entities = []
+    for entity_config in platform_config:
+        entities.append(KNXNumber(xknx, entity_config))
+
+    async_add_entities(entities)
+
+
+def _create_expose_sensor(xknx: XKNX, config: ConfigType) -> NumericValue:
+    """Return a KNX NumericValue to be used within XKNX."""
+    return NumericValue(
+        xknx,
+        name=config[CONF_NAME],
+        group_address=config[KNX_ADDRESS],
+        group_address_state=config.get(CONF_STATE_ADDRESS),
+        respond_to_read=config[CONF_RESPOND_TO_READ],
+        value_type=config[CONF_TYPE],
+    )
+
+
+class KNXNumber(KnxEntity, NumberEntity, RestoreEntity):
+    """Representation of a KNX number."""
+
+    def __init__(self, xknx: XKNX, config: ConfigType) -> None:
+        """Initialize a KNX number."""
+        self._device: NumericValue
+        super().__init__(_create_expose_sensor(xknx, config))
+        self._unique_id = f"{self._device.sensor_value.group_address}"
+
+        self._min_value: float = max(
+            config.get(
+                NumberSchema.CONF_MIN,
+                self._device.sensor_value.dpt_class.value_min,
+            ),
+            self._device.sensor_value.dpt_class.value_min,
+        )
+        self._max_value: float = min(
+            config.get(
+                NumberSchema.CONF_MAX,
+                self._device.sensor_value.dpt_class.value_max,
+            ),
+            self._device.sensor_value.dpt_class.value_max,
+        )
+        self._step: float = max(
+            config.get(
+                NumberSchema.CONF_STEP,
+                self._device.sensor_value.dpt_class.resolution,
+            ),
+            self._device.sensor_value.dpt_class.resolution,
+        )
+        self._device.sensor_value.value = max(0, self._min_value)
+
+    async def async_added_to_hass(self) -> None:
+        """Restore last state."""
+        await super().async_added_to_hass()
+        if not self._device.sensor_value.readable and (
+            last_state := await self.async_get_last_state()
+        ):
+            if last_state.state not in (STATE_UNKNOWN, STATE_UNAVAILABLE):
+                self._device.sensor_value.value = float(last_state.state)
+
+    @property
+    def min_value(self) -> float:
+        """Return the minimum value."""
+        return self._min_value
+
+    @property
+    def max_value(self) -> float:
+        """Return the maximum value."""
+        return self._max_value
+
+    @property
+    def step(self) -> float:
+        """Return the increment/decrement step."""
+        return self._step
+
+    @property
+    def value(self) -> float:
+        """Return the entity value to represent the entity state."""
+        value = self._device.resolve_state()
+        # self._device.sensor_value.value is initialized so this should never be None
+        return value if value is not None else 0
```

Comment:
should I use `assert value is not None` here?

Reply:
No i thing returning None is better. If it's due to failure to communicate, the entity aight to go unavailable

Reply:
Return value of the value property is defined `float`, not `Optional[float]`. 
Should I just type: ignore this?
Communication is not an issue here as it is set in `__init__()` above directly. It's just a type checker issue.

Reply:
so I think the options are
- leave it as is
- `assert value is not None`
- `cast(float, value)`
- `# type: ignore`

Reply:
If you are 100% sure it is always a float that it returns, casting makes sense imho.